index.html

```diff
@@ -2448,6 +2448,128 @@ <h2>
         </section>
       </section>
     </section>
+    <section>
+      <h2>
+        Launch queue and launch params
+      </h2>
+      <p>
+        TODO: This should move to its own spec.
+      </p>
+      <section data-dfn-for="LaunchParams">
+        <h3>
+          <dfn>LaunchParams</dfn> interface
+        </h3>
+        <pre class="idl">
+          [Exposed=Window] interface LaunchParams {
+            readonly attribute DOMString? targetURL;
+          };
+        </pre>
+        <section>
+          <h2>
+            <dfn>targetURL</dfn> attribute
+          </h2>
+          <p>
+            The {{LaunchParams/targetURL}} attribute is a [=string=] that 
+            represents the [=URL=] of the launch [=manifest/within scope=] 
+            of the application. This attribute is set by the 
+            [=Handling Web App Launches=] process.
+          </p>
+        </section>
+      </section>
+      <section data-dfn-for="LaunchConsumer">
+        <h3>
+          <dfn>LaunchConsumer</dfn> function
+        </h3>
+        <pre class="idl">
+          callback LaunchConsumer = any (LaunchParams params);
+        </pre>
+      </section>
+      <section data-dfn-for="LaunchQueue">
+        <h3>
+          <dfn>LaunchQueue</dfn> interface
+        </h3>
+        <pre class="idl">
+          partial interface Window {
+            readonly attribute LaunchQueue launchQueue;
+          };
+
+          [Exposed=Window] interface LaunchQueue {
+            undefined setConsumer(LaunchConsumer consumer);
+          };
+        </pre>
+        <p>
+          {{LaunchQueue}} has an <dfn>unconsumed launch params</dfn> which 
+          is a [=list=] of {{LaunchParams}}. [=Unconsumed launch params=] is 
+          initially empty.
+        </p>
+        <p>
+          {{LaunchQueue}} has an <dfn>assigned launch consumer</dfn>.
+        </p>
+        <section>
+          <h2>
+            <dfn>setConsumer</dfn> method
+          </h2>
+          <p>
+            The {{LaunchQueue/setConsumer(consumer)}} method steps are:
+          </p>
+          <ol class="algorithm">
+            <li>Set the [=assigned launch consumer=] to |consumer|.
+            </li>
+            <li>[=list/For each=] [=unconsumed launch params=]:
+            </li>
+            <li>
+              <ol>
+          <li>Call |consumer| with [=unconsumed launch params=].
+          </li>
+              </ol>
+            </li>
+            <li>Clear the list of [=unconsumed launch params=].
+            </li>
+          </ol>
+        </section>
+      </section>
+      <section>
+        <h3>
+          Handling Web App Launches
+        </h3>
+        <p>
+          When a web is launched, the following steps are run:
```

Comment:
s/web/web app

Reply:
The general idea of "launching" is implicit in the rest of the spec, even in the first sentence! https://www.w3.org/TR/appmanifest/#web-application-manifest

This algorithm should probably be altered somewhat so that "Execute a file handler launch" can reuse it with slight tweaks perhaps by making the LaunchParams an input which, if not specified by the caller because the caller is the hand wavy concept of "launching a web app", defaults to just containing the `start_url`.

Reply:
I think we can probably just delegate the matching of file or file list -> web app as an operation of the operating system. This lets us not worry about the extensions etc - we can just say we register this information with the operating system I think? This allows us to not worry about mime types either.

Then we can start our algorithm from getting triggered by the operating system, where we already have a list of files and the reference to the applicable manifest/web app. We can say that this should be consistent with how any application is triggered when double-clicking a file or list of files in the operating system.

I'm not sure the best way to introduce this - we could define 'launching a web application' as an algorithm in general, which we don't really do yet. Then that can optionally take files?  But I don't think we need to specify an algorithm about how those files are chosen - I think we can just hand-wave at what operating systems do currently (as we would be technically, at that point, writing spec for operating system behavior and not user agents, and I think that is out-of-scope and probably subject to change).

Reply:
> I think we can probably just delegate the matching of file or file list -> web app as an operation of the operating system. This lets us not worry about the extensions etc - we can just say we register this information with the operating system I think? This allows us to not worry about mime types either.

The algorithm I sketched out doesn't match a file [list] to a web app. Given a web app and a file [list], it finds which file handler item in the web app manifest to use, since there could be more than one and each has its own start url. It is true that the above algorithms do not need to worry about mime types; that's solely an OS concern.

> This algorithm should probably be altered somewhat so that "Execute a file handler launch" can reuse it with slight tweaks perhaps by making the LaunchParams an input which, if not specified by the caller because the caller is the hand wavy concept of "launching a web app", defaults to just containing the start_url.

this sg.